botorch/optim/optimize.py

```diff
@@ -257,22 +270,49 @@ def _optimize_acqf_batch(opt_inputs: OptimizeAcqfInputs) -> tuple[Tensor, Tensor
 
     initial_conditions_provided = opt_inputs.batch_initial_conditions is not None
 
+    required_raw_samples = opt_inputs.raw_samples
     if initial_conditions_provided:
-        batch_initial_conditions = opt_inputs.batch_initial_conditions
+        provided_initial_conditions = opt_inputs.batch_initial_conditions
+        if opt_inputs.raw_samples is not None:
+            required_raw_samples -= provided_initial_conditions.shape[-2]
     else:
+        provided_initial_conditions = None
+
+    if required_raw_samples is not None and required_raw_samples > 0:
         # pyre-ignore[28]: Unexpected keyword argument `acq_function` to anonymous call.
-        batch_initial_conditions = opt_inputs.get_ic_generator()(
+        generated_initial_conditions = opt_inputs.get_ic_generator()(
             acq_function=opt_inputs.acq_function,
             bounds=opt_inputs.bounds,
             q=opt_inputs.q,
             num_restarts=opt_inputs.num_restarts,
-            raw_samples=opt_inputs.raw_samples,
+            raw_samples=required_raw_samples,
             fixed_features=opt_inputs.fixed_features,
             options=options,
             inequality_constraints=opt_inputs.inequality_constraints,
             equality_constraints=opt_inputs.equality_constraints,
             **opt_inputs.ic_gen_kwargs,
         )
+    else:
+        generated_initial_conditions = None
+
+    if provided_initial_conditions is not None and generated_initial_conditions is not None:
+        if (  # Repeat the provided initial conditions to match the number of restarts
+            provided_initial_conditions.shape[0] == 1
+            and opt_inputs.num_restarts is not None
+            and opt_inputs.num_restarts > 1
+        ):
+            provided_initial_conditions = provided_initial_conditions.repeat(
+                opt_inputs.num_restarts, *([1] * (provided_initial_conditions.dim()-1))
+            )
+        batch_initial_conditions = torch.cat(
+            [provided_initial_conditions, generated_initial_conditions], dim=-2
+        )  # should this be shuffled?
+    elif provided_initial_conditions is not None:
+        batch_initial_conditions = provided_initial_conditions
+    elif generated_initial_conditions is not None:
+        batch_initial_conditions = generated_initial_conditions
+    else:
+        raise ValueError("Either `batch_initial_conditions` or `raw_samples` must be set.")
```

Comment:
This has some  nontrivial complexity and makes it more challenging to read - could you factor this out into some helper function?